airflow/jobs/scheduler_job_runner.py

```diff
@@ -1838,12 +1890,63 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
     def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[TaskInstance]]:
         """Organize TIs into lists per their respective executor."""
         _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
-        executor: str | None
         for ti in tis:
-            if ti.executor:
-                executor = str(ti.executor)
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                _executor_to_tis[executor_obj].append(ti)
             else:
-                executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            if TYPE_CHECKING:
+                # All executors should have a name if they are initted from the executor_loader. But we need
+                # to check for None to make mypy happy.
+                assert executor.name
+            executor_to_slots_available[executor.name] = executor.slots_available
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                ti_exec_name = executor_obj.name
+                if TYPE_CHECKING:
+                    # All executors should have a name if they are initted from the executor_loader. But we
+                    # need to check for None to make mypy happy.
+                    assert ti_exec_name
+                if executor_to_slots_available[ti_exec_name] > 0:
+                    tis_we_have_room_for.add(ti)
+                    executor_to_slots_available[ti_exec_name] -= 1
+            else:
+                continue
+
+        return tis_we_have_room_for
+
+    def _try_to_load_executor(self, executor_name: str | Column | None) -> BaseExecutor | None:
+        """Try to load the given executor.
+
+        In this context, we don't want to fail if the executor does not exist. Catch the exception and
+        log to the user.
+        """
+        # Executor name comes in as a column type which upsets mypy. Here we are "casting" the value to
+        # either None or a string to satisfy typing
+        if not executor_name:
+            executor_name = None
+        else:
+            executor_name = str(executor_name)
+        try:
+            return ExecutorLoader.load_executor(executor_name)
+        except AirflowException as e:
+            if "Unknown executor" in str(e):
```

Comment:
@o-nikolas i am not seeing anywhere that we raise an error that has `Unknown executor` in the name.  am i missing something?

I see this is the message that is potentially raised in `load_executor`
```
            raise AirflowConfigException(
                f'The module/attribute could not be loaded. Please check "executor" key in "core" section. '
                f'Current value: "{_executor_name}".'
            )
```

Reply:
That one is thrown by the `lookup_executor_name_by_str` method [here](https://github.com/apache/airflow/blob/5387ec104234e3e9e79f3048c1d8121ee99c8e1e/airflow/executors/executor_loader.py#L198)

I'm working on the backfill changes now and need to handle that error as well (and we also do the same in the dag parsing logic that asserts the executor requested is available). So I think I'll actually change it to a custom exception at this point, instead of catching ValueError and inspecting the message which can be a bit brittle. Hopefully, it'll be easier to trace the origin that way too!

Reply:
This case in general _should_ be very unlikey, since the dag parser should catch those earlier. But I thought I'd put a guard here just in case, since it's pretty critical if you do somehow arrive here and the user should get feedback about that.

Reply:
oh yeah `lookup_executor_name_by_str` just found it thanksn

Reply:
i think custom exception is good idear 👍